md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2370 @@
+
+The requirements, plus the various minor requirements, can be validated in a
+single linear pass as follows:
+
+The following data structures are created:
+ - A *type stack*, with entries each containing a [type], for tracking the
+   sequence of values that will be on the value stack at execution time.
+ - A *control-flow stack*, with entries each containing
+    - A *limit* integer value, which is an index into the type stack indicating
+      the boundary between values pushed before the current region and values
+      pushed inside the current region.
+    - An optional [type] sequence, which is used to check that all exits from
+      the current region leave the same sequence of types on the type stack.
+
+The following invariants are required to be preserved:
+ - Whenever an element of either stack is to be popped, that stack is required
+   to be non-empty.
+ - Whenever an element of either stack is to be accessed by index, the index is
+   required to be within the bounds of that stack.
+ - When a [type] is to be popped from the type stack, the length of the type
+   stack is required to be greater than the control-flow stack top's limit
+   value.
+
+The type stack begins empty. The control-flow stack begins with one entry, with
+the [type] sequence consisting of the return types of the function, and with the
+limit value being zero.
+
+If the function's instruction sequence is empty, or if the last instruction in
+the sequence isn't an [`end`](#end), validation fails.
+
+For each instruction in the body, in sequence, if the instruction has a
+**Validation** clause, it is validated according to that clause, otherwise
+[generic instruction validation](#generic-instruction-validation) is performed.
+
+Finally, for each type in the function's return list, a type is popped from the
+type stack and required be the same. The type stack and the control-flow stack
+are then both required to be empty. If no failures were detected anywhere in the
+function, function validation is successful.
+
+> The control-flow stack's limit values effectively mark region boundaries in
+the type stack. Regions are required to be nested, and each region's limit value
+is greater than those of the regions that enclose it. Things pushed onto the
+type stack outside a region cannot be popped from within the region.
+
+> The final [`end`](#end) instruction may be implicit in some representations.
+
+TODO: Will the `end` be made explicit? Monitor
+https://github.com/WebAssembly/design/pull/666.
+
+#### Generic Instruction Validation
+
+For each operand [type] in the instruction's signature in reverse order, a
+[type] is popped from the type stack and required to match it. Each return type
+of the instruction's signature is then pushed onto the type stack.
+
+If the instruction's signature has no return types:
+ - If the length of the type stack is greater than the control-flow stack top's
+   limit value, validation fails.
+
```

Comment:
Functions with multiple return values are not permitted in this version of the spec. If they are added, I agree that they would want to be accompanied by a mechanism to access their return values in a flexible way.